velox/exec/SortBuffer.h

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <vector/BaseVector.h>
+#include "velox/exec/ContainerRowSerde.h"
+#include "velox/exec/Operator.h"
+#include "velox/exec/OperatorUtils.h"
+#include "velox/exec/RowContainer.h"
+#include "velox/exec/Spill.h"
+
+namespace facebook::velox::exec {
+
+/// A utility class to accumulate data inside and output the sorted result.
+/// Spilling would be triggered if spilling is enabled and memory usage exceeds
+/// limit.
+class SortBuffer {
+ public:
+  SortBuffer(
+      const RowTypePtr& input,
+      const std::vector<column_index_t>& sortedIndex,
+      const std::vector<CompareFlags>& sortedProperties,
+      velox::memory::MemoryPool* pool,
+      tsan_atomic<bool>* nonReclaimableSection,
+      uint32_t outputBatchSizeLimit_ = 10000,
+      std::optional<Spiller::Config> spillConfig = std::nullopt,
+      uint64_t spillMemoryThreshold = 0);
+
+  void addInput(const RowVectorPtr& input);
+
+  /// Indicates no more input. After it's been called, addInput() can't be
+  /// called on SortBuffer. This function should only be called once as it would
+  /// trigger either of:
+  ///  - internal sorting
+  ///  - finish the spilling and setup merge reader to prepare un-spilling
+  std::optional<SpillStats> noMoreInput();
+
+  /// We break the result into batches, each call of getOutput() return one
+  /// batch of result, output() returning nullptr indicates all results have
+  /// been returned.
+  RowVectorPtr getOutput();
+
+  /// API for Abitrator to recycle memory from this object.
+  /// Will trigger spilling and free targetBytes of memory.
+  void reclaim(uint64_t targetBytes);
+
+ private:
+  /// Trigger disk spilling if needed. There are a few scenarios that would
+  /// trigger spilling (when spilling is enabled):
+  ///  - current memory usage exceeds spilling threshold
+  ///  - current available reservation in memory pool is not enough and
+  ///  furthermore we fail to reserve more in memory pool
+  void ensureInputFits(const RowVectorPtr& input);
+  /// Spill targetRows/targetBytes from 'data_' to disk to free some memory.
+  void spill(int64_t targetRows, int64_t targetBytes);
+  /// Initialize the reusable output buffer, if it's initialized already, reset
+  /// it for next iteration.
+  void initOrResetOutputBuffer();
```

Comment:
Please keeps the original method name in order by. We just need to highlight what a function does.